### PR TITLE
add support for alternate json Content-Types

### PR DESCRIPTION
**Pull request by pfefferle**
Add support for alternate JSON Content-Types like `application/ld+json`.

* List of Media Types: https://www.iana.org/assignments/media-types/media-types.xhtml#application
* Media-Type Naming Requirements: https://tools.ietf.org/html/rfc6838#section-4.2

Trac ticket: https://core.trac.wordpress.org/ticket/49404

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
Thanks for the PR @pfefferle! 

Why is `get_parameter_order` using a different check from `parse_json_params`? 

We'll probably also want to update `wp_is_json_request`.

**Comment by pfefferle** [comment]
I thought it might be more performat to only check if there is `JSON` in the content type, because otherwise `preg_match` would be called twice. But sure, I can use the `preg_match` in both cases.

**Comment by TimothyBJacobs** [comment]
My concern about making the parameter order check looser is that we'd be saying this request contains JSON data when really it doesn't.

Maybe introduce a protected `is_json_request()` method on the request object and cache the results of the regex check?

**Comment by pfefferle** [comment]
@TimothyBJacobs what do you think about the `preg_match` in general?

**Comment by TimothyBJacobs** [comment]
I like it!

**Comment by pfefferle** [comment]
I like the `is_json_request()` idea, but I think `is_json_content_type()` would be a better match, because it only checks the content_type and the content body still might be broken/non JSON?!?

**Comment by TimothyBJacobs** [comment]
That works for me!

**Comment by pfefferle** [comment]
@TimothyBJacobs I added a `is_json_content_type` method with some tests.

**Comment by TimothyBJacobs** [comment]
@pfefferle can we cache that value as long as the `content-type` header hasn't changed? That way we don't have to worry as much about the impact of the regex check?

**Comment by pfefferle** [comment]
@TimothyBJacobs oops, forgot about that!

**Comment by TimothyBJacobs** [comment]
We also have to worry about `add_header` and `remove_header`. It may be better to build into the cache the value of the content-type header when it evaluated it.

**Comment by pfefferle** [comment]
@TimothyBJacobs do you think it is worth it? I do think the complete "caching" and "invalidate caches" is very error-prone and it is "only" a `preg_match` in the end?!?

**Comment by TimothyBJacobs** [comment]
You're definitely right that is error prone. If the `preg_match` would only occur a couple of times in a request, I think it wouldn't be too bad. But AFAICT, this would result in a `preg_match` call for every single get or set of a parameter which could add up.

**Comment by pfefferle** [comment]
@TimothyBJacobs like this? we should discuss the namings!

**Comment by TimothyBJacobs** [comment]
I like that! Let's get some tests on the different ways we can modify the header.

> we should discuss the namings!

Namings make sense to me! Do you have any concerns about them?

**Comment by pfefferle** [comment]
@TimothyBJacobs while writing tests, I think I found a bug. `Content-Type` should always contain a single type, even if you use `$request->add_header();`, but it works like the `Accept` header and adds them to a list.

```PHP
$request->add_header( 'Content-Type', 'application/json' );
$request->add_header( 'Content-Type', 'application/activity+json' );
```

Result for `$request->get_header( 'Content-Type' )` is `application/json,application/activity+json`

Result for `$request->get_content_type()` is:

```
array(4) {
  ["value"]=>
  string(78) "application/json,application/activity+json"
  ["type"]=>
  string(11) "application"
  ["subtype"]=>
  string(66) "json,application/activity+json"
  ["parameters"]=>
  string(0) ""
}
```

**Comment by TimothyBJacobs** [comment]
I think we should consider that in a separate ticket. For now let's just keep operating on the first header in the list.

**Comment by pfefferle** [comment]
I only mentioned it, because it broke the test for `add_header` but I will rewrite them.

**Comment by pfefferle** [comment]
`wp_is_json_request` should use the same `preg_match` pattern https://wordpress.slack.com/archives/C02RQC26G/p1591294944063300

**Comment by pfefferle** [comment]
@TimothyBJacobs I added tests. Can you re-review?

**Comment by TimothyBJacobs** [comment]
Do you want to update `wp_is_json_request` in this PR as well? Would probably be good to keep them synced like you mentioned.

**Comment by pfefferle** [comment]
Sure!

**Comment by TimothyBJacobs** [comment]
@pfefferle following up on this. If we want to include it in 5.6 I'd like to have a patch in the next few days. Ideally by Monday.

**Comment by TimothyBJacobs** [comment]
@pfefferle What about localizing the cache to `wp_is_json_content_type()` with a static variable? That way we can simplify the changes to `WP_REST_Request`, and would make caching `wp_is_json_request()` simple as well.

**Comment by pfefferle** [comment]
@TimothyBJacobs can you re-check?

**Comment by TimothyBJacobs** [comment]
@pfefferle This is looking good, thanks! To clarify, since `wp_is_json_request` can now be used by multiple consumers, it should cache any input, not just the last one. So a `static $is_json[ $header ] = regex_check()`

File: src/wp-includes/load.php [hunk 1]
```diff
@@ -1526,6 +1526,30 @@ function wp_is_jsonp_request() {
...
+ *
+ * @since 5.6.0
+ *
+ * @staticvar bool $is_json
+ *
+ * @param string $content_type A Content-Type string
+ * @return bool True if string is a valid JSON Content-Type.
+ */
+function wp_is_json_content_type( $content_type ) {
+	static $is_json_content_type = array();
```
Comment (TimothyBJacobs) [inline src/wp-includes/load.php:1540]: Let's shorten this to just `$cache`.

File: src/wp-includes/load.php [hunk 2]
```diff
@@ -1526,6 +1526,30 @@ function wp_is_jsonp_request() {
...
+
+	if ( ! isset( $is_json_content_type[ $content_type ] ) ) {
+		$is_json_content_type[ $content_type ] = (bool) preg_match( '/(^|\s|,)application\/([\w!#\$&-\^\.\+]+\+)?json(\+oembed)?($|\s|;|,)/i', $content_type );
+	}
+
+	if ( ! is_bool( $is_json_content_type[ $content_type ] ) ) {
```
Comment (TimothyBJacobs) [inline src/wp-includes/load.php:1546]: How would this branch get executed? The result of `preg_match` is cased to a boolean.

File: src/wp-includes/rest-api/class-wp-rest-request.php [hunk 3]
```diff
@@ -127,6 +127,23 @@ public function __construct( $method = '', $route = '', $attributes = array() )
...
+	 * @return boolean
+	 */
+	public function is_json_content_type() {
+		$content_type = $this->get_content_type();
+
+		if ( isset( $content_type['value'] ) && wp_is_json_content_type( $content_type['value'] ) ) {
+			return true;
+		}
+
+		return false;
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-request.php:144]:
```suggestion
		return isset( $content_type['value'] ) && wp_is_json_content_type( $content_type['value'] )
```

File: tests/phpunit/tests/functions.php [hunk 4]
```diff
@@ -1743,4 +1743,26 @@ public function data_test_human_readable_duration() {
 			array( '03:61:59', false ), // Out of bound.
 		);
 	}
+
+	/**
+	 * @dataProvider data_test_wp_is_json_content_type
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/functions.php:1748]: Needs an `@ticket` annotation

File: tests/phpunit/tests/rest-api/rest-request.php [hunk 5]
```diff
@@ -178,6 +178,82 @@ public function test_parameter_order_post() {
...
+			array( 'application/ld+json; profile="https://www.w3.org/ns/activitystreams"', 'json', true ),
+			array( 'application/activity+json', 'json', true ),
+			array( 'application/json+oembed', 'json', true ),
+			array( 'application/nojson', 'body', false ),
+			array( 'application/no.json', 'body', false ),
+		);
+	}
+
+	/**
+	 * @dataProvider alternate_json_content_type_provider
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-request.php:193]: Needs an `@ticket` annotation.

File: tests/phpunit/tests/rest-api/rest-request.php [hunk 6]
```diff
@@ -178,6 +178,82 @@ public function test_parameter_order_post() {
...
+			array( 'application/ld+json; profile="https://www.w3.org/ns/activitystreams"', true ),
+			array( 'application/activity+json', true ),
+			array( 'application/json+oembed', true ),
+			array( 'application/nojson', false ),
+			array( 'application/no.json', false ),
+		);
+	}
+
+	/**
+	 * @dataProvider is_json_content_type_provider
```
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-request.php:223]: Needs an `@ticket` annotation.

File: src/wp-includes/load.php (hunk 2 above)
Comment (pfefferle) [reply src/wp-includes/load.php:1546]: It returns `1` if there is a match, `0` if there is none and `false` if there is an error, so casting to bool, should be fine.

Comment (TimothyBJacobs) [reply src/wp-includes/load.php]: I agree casting to book is fine, but that means that the `is_bool` check should never return false, no?

Comment (pfefferle) [reply src/wp-includes/load.php]: I agree, I removed it in one of my last commits.

File: src/wp-includes/rest-api/class-wp-rest-request.php [hunk 7]
```diff
@@ -127,6 +127,19 @@ public function __construct( $method = '', $route = '', $attributes = array() )
 		$this->set_attributes( $attributes );
 	}
 
+	/**
+	 * Checks if Content-Type is JSON.
+	 *
+	 * @since @todo
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-request.php:133]:
```suggestion
	 * @since 5.6.0
```

File: src/wp-includes/rest-api/class-wp-rest-request.php [hunk 8]
```diff
@@ -127,6 +127,19 @@ public function __construct( $method = '', $route = '', $attributes = array() )
 		$this->set_attributes( $attributes );
 	}
 
+	/**
+	 * Checks if Content-Type is JSON.
+	 *
+	 * @since 5.6.0
+	 *
+	 * @return boolean
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-request.php:135]:
```suggestion
	 * @return bool
```

File: src/wp-includes/rest-api/class-wp-rest-request.php [hunk 9]
```diff
@@ -127,6 +127,19 @@ public function __construct( $method = '', $route = '', $attributes = array() )
 		$this->set_attributes( $attributes );
 	}
 
+	/**
+	 * Checks if Content-Type is JSON.
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-request.php:131]:
```suggestion
	 * Checks if the request's Content-Type is JSON.
```

**Comment by pfefferle** [comment]
👍

**Comment by TimothyBJacobs** [comment]
Fixed in d56906f5b0fec8cce8829bb548a8c7c7e4c64b80. Thanks for the patch @pfefferle!
